ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/type/SszPublicKey.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.datastructures.type;
+
+import org.apache.tuweni.bytes.Bytes48;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.ssz.backing.collections.impl.SszByteVectorImpl;
+import tech.pegasys.teku.ssz.backing.tree.TreeNode;
+
+public class SszPublicKey extends SszByteVectorImpl {
+
+  private BLSPublicKey publicKey;
+
+  public SszPublicKey(Bytes48 publicKeyBytes) {
+    super(SszPublicKeySchema.INSTANCE, publicKeyBytes);
+  }
+
+  public SszPublicKey(BLSPublicKey publicKey) {
+    super(SszPublicKeySchema.INSTANCE, publicKey.toBytesCompressed());
+    this.publicKey = publicKey;
+  }
+
+  SszPublicKey(TreeNode backingNode) {
+    super(SszPublicKeySchema.INSTANCE, backingNode);
+  }
+
+  public BLSPublicKey getBLSPublicKey() {
+    if (publicKey == null) {
+      publicKey = BLSPublicKey.fromBytesCompressed(getBytes());
+    }
+    return publicKey;
```

Comment:
I worry that this isn't thread safe. We often access states and blocks from multiple threads because they're shared via the cache so I suspect we need thread safety here.

`Suppliers.memoize` is very useful for this kind of thing.

Reply:
Yeah, I also love `Suppliers.memoize` but unfortunately they shown poor performance in critical places and replacing them with similar constructs yielded performance gain. 
I think, with this construct the maximum 'harm' would be creating `BLSPublicKey.fromBytesCompressed(getBytes())` more that once if accessed in parallel for the first time. 

I replaced the field with `Suppliers.memoize` for now (05539c52). We could always roll this back in case of performance issue

Reply:
If `Suppliers.memoize` is too expensive we could make the field `volatile` which would provide enough safety without locks (but goes back to possibly initing it multiple times in parallel).

I'd probably have to dig into the java memory model to be sure, but unless the field is volatile, there's no guarantee that the construction of `BLSPublicKey` is completed prior to assigning to the variable (compilers or CPUs can reorder and memory visibility isn't guaranteed).  So the worst case is that another thread sees a partially initialised `BLSPublicKey` instance.

The piece I'm not 100% sure of is that there are some extra guarantees about visibility when creating classes that only have `final` fields which might make it safe without needing `Suppliers.memoize` or `volatile`.

Reply:
Yeah, you are right. I was missing that 'partial inited instance' case here. 

Though this particular case seems to be safe according to JMM: the `BLSPublicKey` has 2 `final Supplier` fields (`Supplier` is thread-safe itself). It may become unsafe if `BLSPublicKey` construction changed at some point, so it's always better to be secured with `volatile`